### PR TITLE
🗺️ Atlas: Refactor core/vector.rs to modular structure

### PR DESCRIPTION
**Pull request by madmax983**
🕸️ Tangle: `src/core/vector.rs` was a 5300+ line "Blob" mixing types, SIMD internals, ops, and tests.
📐 Blueprint: Split into `src/core/vector/` with `types.rs`, `sparse.rs`, `simd.rs`, `ops.rs`, `validation.rs`, `metric.rs`, and `tests.rs`.
🧱 Stability: Strict separation of concerns. `mod.rs` preserves public API.
🔬 Verification: Unit tests passed (`cargo test core::vector`). Clippy passed.

---
*PR created automatically by Jules for task [11861150725542821329](https://jules.google.com/task/11861150725542821329) started by @madmax983*

### CODE REVIEW

File: src/core/vector/simd.rs [hunk 1]
```diff
@@ -0,0 +1,615 @@
...
+            for i in 0..remainder {
+                let ai = a[start + i];
+                let bi = b[start + i];
+                dot_rem += ai * bi;
+                mag_a_rem += ai * ai;
+                mag_b_rem += bi * bi;
+            }
+
+            (dot + dot_rem, mag_a + mag_a_rem, mag_b + mag_b_rem)
+        }
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/core/vector/simd.rs:66]: > medium
> 
> For better readability and safety, this function can be refactored to use `chunks_exact` and iterators for processing the vector chunks and their remainder. This approach is more idiomatic than manual pointer arithmetic and indexing, and it's consistent with how other functions like `dot_product_avx2` are implemented in this file.
> 
> ```rust
>         let a_chunks = a.chunks_exact(8);
>         let b_chunks = b.chunks_exact(8);
>         let a_rem = a_chunks.remainder();
>         let b_rem = b_chunks.remainder();
> 
>         // Accumulators for 8 floats at a time
>         let mut dot_acc = _mm256_setzero_ps();
>         let mut mag_a_acc = _mm256_setzero_ps();
>         let mut mag_b_acc = _mm256_setzero_ps();
> 
>         // Process 8 floats at a time
>         for (va_chunk, vb_chunk) in a_chunks.zip(b_chunks) {
>             let va = _mm256_loadu_ps(va_chunk.as_ptr());
>             let vb = _mm256_loadu_ps(vb_chunk.as_ptr());
> 
>             // Fused multiply-add for dot product and magnitudes
>             dot_acc = _mm256_fmadd_ps(va, vb, dot_acc);
>             mag_a_acc = _mm256_fmadd_ps(va, va, mag_a_acc);
>             mag_b_acc = _mm256_fmadd_ps(vb, vb, mag_b_acc);
>         }
> 
>         // Horizontal sum of 256-bit vectors
>         let dot = horizontal_sum_avx(dot_acc);
>         let mag_a = horizontal_sum_avx(mag_a_acc);
>         let mag_b = horizontal_sum_avx(mag_b_acc);
> 
>         // Handle remainder with safe scalar operations.
>         let (dot_rem, mag_a_rem, mag_b_rem) =
>             a_rem.iter().zip(b_rem.iter()).fold(
>                 (0.0f32, 0.0f32, 0.0f32),
>                 |(dot, mag_a, mag_b), (&ai, &bi)| {
>                     (dot + ai * bi, mag_a + ai * ai, mag_b + bi * bi)
>                 },
>             );
> 
>         (dot + dot_rem, mag_a + mag_a_rem, mag_b + mag_b_rem)
> ```

@jules

File: src/core/vector/simd.rs [hunk 2]
```diff
@@ -0,0 +1,615 @@
...
+            let start = chunks * 4;
+            for i in 0..remainder {
+                let ai = a[start + i];
+                let bi = b[start + i];
+                dot_rem += ai * bi;
+                mag_a_rem += ai * ai;
+                mag_b_rem += bi * bi;
+            }
+
+            (dot + dot_rem, mag_a + mag_a_rem, mag_b + mag_b_rem)
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/core/vector/simd.rs:126]: > medium
> 
> Similar to `dot_and_magnitudes_avx2`, this function can be improved by using `chunks_exact` and iterators instead of manual indexing. This makes the code more idiomatic and safer.
> 
> ```rust
>         let a_chunks = a.chunks_exact(4);
>         let b_chunks = b.chunks_exact(4);
>         let a_rem = a_chunks.remainder();
>         let b_rem = b_chunks.remainder();
> 
>         // Accumulators for 4 floats at a time
>         let mut dot_acc = _mm_setzero_ps();
>         let mut mag_a_acc = _mm_setzero_ps();
>         let mut mag_b_acc = _mm_setzero_ps();
> 
>         // Process 4 floats at a time
>         for (va_chunk, vb_chunk) in a_chunks.zip(b_chunks) {
>             let va = _mm_loadu_ps(va_chunk.as_ptr());
>             let vb = _mm_loadu_ps(vb_chunk.as_ptr());
> 
>             // Multiply and accumulate
>             dot_acc = _mm_add_ps(dot_acc, _mm_mul_ps(va, vb));
>             mag_a_acc = _mm_add_ps(mag_a_acc, _mm_mul_ps(va, va));
>             mag_b_acc = _mm_add_ps(mag_b_acc, _mm_mul_ps(vb, vb));
>         }
> 
>         // Horizontal sum of 128-bit vectors
>         let dot = horizontal_sum_sse(dot_acc);
>         let mag_a = horizontal_sum_sse(mag_a_acc);
>         let mag_b = horizontal_sum_sse(mag_b_acc);
> 
>         // Handle remainder with safe scalar operations.
>         let (dot_rem, mag_a_rem, mag_b_rem) =
>             a_rem.iter().zip(b_rem.iter()).fold(
>                 (0.0f32, 0.0f32, 0.0f32),
>                 |(dot, mag_a, mag_b), (&ai, &bi)| {
>                     (dot + ai * bi, mag_a + ai * ai, mag_b + bi * bi)
>                 },
>             );
> 
>         (dot + dot_rem, mag_a + mag_a_rem, mag_b + mag_b_rem)
> ```

@jules

File: src/core/vector/simd.rs [hunk 3]
```diff
@@ -0,0 +1,615 @@
...
+            let mut sum = horizontal_sum_avx(acc);
+
+            // Handle remainder with scalar operations
+            let start = chunks * 8;
+            for i in 0..remainder {
+                let diff = a[start + i] - b[start + i];
+                sum += diff * diff;
+            }
+
+            sum
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/core/vector/simd.rs:274]: > medium
> 
> This function can be refactored to use `chunks_exact` and iterators for processing the vector chunks and their remainder. This approach is more idiomatic and safer than manual pointer arithmetic and indexing.
> 
> ```rust
>         // SAFETY: The unsafe block is required by the `unsafe_op_in_unsafe_fn` lint.
>         // All unsafe operations within this unsafe fn must still be in an unsafe block.
>         // The caller guarantees AVX2 and FMA are available via runtime feature detection.
>         let a_chunks = a.chunks_exact(8);
>         let b_chunks = b.chunks_exact(8);
>         let a_rem = a_chunks.remainder();
>         let b_rem = b_chunks.remainder();
> 
>         // Accumulator for 8 floats at a time
>         let mut acc = _mm256_setzero_ps();
> 
>         // Process 8 floats at a time
>         for (va_chunk, vb_chunk) in a_chunks.zip(b_chunks) {
>             let va = _mm256_loadu_ps(va_chunk.as_ptr());
>             let vb = _mm256_loadu_ps(vb_chunk.as_ptr());
> 
>             // Compute difference
>             let diff = _mm256_sub_ps(va, vb);
> 
>             // Square and accumulate using FMA: acc = diff * diff + acc
>             acc = _mm256_fmadd_ps(diff, diff, acc);
>         }
> 
>         // Horizontal sum of 256-bit vector
>         let mut sum = horizontal_sum_avx(acc);
> 
>         // Handle remainder with scalar operations
>         for (&ai, &bi) in a_rem.iter().zip(b_rem.iter()) {
>             let diff = ai - bi;
>             sum += diff * diff;
>         }
> 
>         sum
> ```

@jules

File: src/core/vector/simd.rs [hunk 4]
```diff
@@ -0,0 +1,615 @@
...
+            let mut sum = horizontal_sum_sse(acc);
+
+            // Handle remainder with scalar operations
+            let start = chunks * 4;
+            for i in 0..remainder {
+                let diff = a[start + i] - b[start + i];
+                sum += diff * diff;
+            }
+
+            sum
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/core/vector/simd.rs:320]: > medium
> 
> This function can be improved by using `chunks_exact` and iterators instead of manual indexing, which is more idiomatic and safer.
> 
> ```rust
>         // SAFETY: The unsafe block is required by the `unsafe_op_in_unsafe_fn` lint.
>         // All unsafe operations within this unsafe fn must still be in an unsafe block.
>         // The caller guarantees SSE2 is available via runtime feature detection.
>         let a_chunks = a.chunks_exact(4);
>         let b_chunks = b.chunks_exact(4);
>         let a_rem = a_chunks.remainder();
>         let b_rem = b_chunks.remainder();
> 
>         // Accumulator for 4 floats at a time
>         let mut acc = _mm_setzero_ps();
> 
>         // Process 4 floats at a time
>         for (va_chunk, vb_chunk) in a_chunks.zip(b_chunks) {
>             let va = _mm_loadu_ps(va_chunk.as_ptr());
>             let vb = _mm_loadu_ps(vb_chunk.as_ptr());
> 
>             // Compute difference
>             let diff = _mm_sub_ps(va, vb);
> 
>             // Square and accumulate
>             acc = _mm_add_ps(acc, _mm_mul_ps(diff, diff));
>         }
> 
>         // Horizontal sum of 128-bit vector
>         let mut sum = horizontal_sum_sse(acc);
> 
>         // Handle remainder with scalar operations
>         for (&ai, &bi) in a_rem.iter().zip(b_rem.iter()) {
>             let diff = ai - bi;
>             sum += diff * diff;
>         }
> 
>         sum
> ```

@jules

**Comment by madmax983** [comment]
> ## Code Review: PR #669 - Vector Module Refactoring
> ### Overview ✅
> This PR successfully refactors the monolithic 5300+ line `src/core/vector.rs` file into a well-organized modular structure. The refactoring maintains the public API while significantly improving code organization and maintainability.
> 
> ## 1. Code Quality & Standards ✅
> **Strengths:**
> 
> * ✅ No `unwrap()` or `expect()` in production code (only in tests/docs)
> * ✅ Strong typing with `VectorDimension` newtype wrapper
> * ✅ Comprehensive error handling with `Result<T>`
> * ✅ Proper input validation throughout
> * ✅ Excellent documentation with examples and performance notes
> 
> **Adherence to CODING_STANDARDS.md:** Exemplary
> 
> ## 2. Bugs & Logic Errors ⚠️
> ### 🔴 CRITICAL: Sparse Vector Magnitude Comparison Bug
> **File:** `src/core/vector/sparse.rs:491`
> 
> ```rust
> if mag_a < SQUARED_MAGNITUDE_THRESHOLD || mag_b < SQUARED_MAGNITUDE_THRESHOLD {
>     return Ok(0.0);
> }
> ```
> 
> **Problem:** Comparing `mag_a` (magnitude) against `SQUARED_MAGNITUDE_THRESHOLD`. Should compare squared magnitudes.
> 
> **Fix:**
> 
> ```rust
> if mag_a * mag_a < SQUARED_MAGNITUDE_THRESHOLD || mag_b * mag_b < SQUARED_MAGNITUDE_THRESHOLD {
>     return Ok(0.0);
> }
> ```
> 
> Or better, use the squared magnitude values directly for the check.
> 
> ### 🟡 MEDIUM: Misleading NaN Count
> **File:** `src/core/vector/sparse.rs:151`
> 
> ```rust
> if val.is_nan() {
>     return Err(VectorError::ContainsNaN { count: 1 }.into());
> }
> ```
> 
> **Problem:** Always reports `count: 1` even if multiple NaN values exist.
> 
> **Recommendation:** Either:
> 
> 1. Count all NaN/Inf values before reporting (requires full scan)
> 2. Report the index of the first invalid value instead of count
> 3. Document that only the first occurrence is reported
> 
> ### 🟢 MINOR: Cosine Distance Range Consideration
> **File:** `src/core/vector/metric.rs:153`
> 
> ```rust
> DistanceMetric::Cosine => cosine_similarity(a, b).map(|sim| 1.0 - sim),
> ```
> 
> **Observation:** Cosine similarity `[-1, 1]` → distance `[0, 2]`. This is documented correctly but could be confusing. Consider whether `(1.0 - sim) / 2.0` for `[0, 1]` range would be more intuitive for distance comparisons. Current implementation is correct, just a UX consideration.
> 
> ### 🟢 MINOR: Documentation Link
> **File:** `src/core/vector/ops.rs:568`
> 
> ```rust
> /// - Dimension limits are enforced at storage time (see [`PropertyValue::vector`])
> ```
> 
> Should be:
> 
> ```rust
> /// - Dimension limits are enforced at storage time (see [`crate::core::PropertyValue::Vector`])
> ```
> 
> ## 3. Performance ✅
> **SIMD Implementation:** Excellent
> 
> * ✅ Runtime feature detection (AVX2/FMA/SSE2)
> * ✅ Proper fallback to scalar on non-x86
> * ✅ FMA usage for AVX2 paths
> * ✅ Efficient remainder handling
> 
> **Allocation Patterns:** Good
> 
> * ✅ `normalize_in_place()` avoids allocation
> * ✅ Sparse operations avoid dense conversions
> * ✅ String allocations only in error paths
> 
> **Recommendation:** Add benchmarks to verify no performance regression from the refactoring.
> 
> ## 4. Security ✅
> **DoS Prevention:** Proper
> 
> * ✅ Dimension validation against `MAX_VECTOR_DIMENSIONS` (100,000)
> * ✅ Index bounds checking in `SparseVec::new()`
> * ✅ No integer overflow opportunities
> * ✅ NaN/Infinity validation
> 
> **SAFETY Comments:** All `unsafe` code has proper `// SAFETY:` comments explaining invariants.
> 
> ## 5. Test Coverage ✅
> The `tests.rs` file (2787 lines) covers:
> 
> * ✅ `VectorDimension` type operations
> * ✅ Edge cases (identical, opposite, orthogonal, zero vectors)
> * ✅ All distance metrics
> * ✅ Normalization (including near-zero vectors)
> * ✅ Sparse vector operations
> * ✅ Dimension mismatches
> * ✅ NaN/Infinity handling
> 
> **Recommendation:** Verify coverage meets 85% threshold with `just coverage-check`.
> 
> ## 6. Module Organization ✅
> **Structure:** Excellent separation of concerns
> 
> ```
> src/core/vector/
> ├── mod.rs          - Module documentation and re-exports
> ├── types.rs        - VectorDimension and constants
> ├── validation.rs   - Input validation functions
> ├── simd.rs        - SIMD implementations (pub(crate))
> ├── ops.rs         - High-level operations
> ├── sparse.rs      - Sparse vector type
> ├── metric.rs      - DistanceMetric enum
> └── tests.rs       - Comprehensive test suite
> ```
> 
> **API Stability:** Public API preserved via re-exports in `mod.rs`.
> 
> ## 7. Documentation Quality ✅
> **Exemplary:**
> 
> * ✅ Module-level documentation
> * ✅ Every public function has rustdoc
> * ✅ Performance characteristics documented
> * ✅ Compiling examples
> * ✅ Edge cases documented
> * ✅ Clear comparison tables
> * ✅ Mathematical formulas
> 
> ## Summary
> **Overall Assessment:** ⭐⭐⭐⭐☆ (4/5 - Excellent with minor fixes needed)
> 
> This is a **high-quality refactoring** that successfully modularizes a large file while maintaining API compatibility. The code follows Rust best practices and has excellent documentation.
> 
> ### 🔴 Required Fixes Before Merge:
> 1. **Fix the sparse vector magnitude comparison bug** (`sparse.rs:491`)
> 2. **Fix or clarify the NaN count reporting** in `SparseVec::new`
> 
> ### 🟡 Recommended Before Merge:
> 1. Run pre-commit checks: `just pre-commit`
> 2. Verify test coverage: `just coverage-check`
> 3. Fix the `PropertyValue::vector` doc link
> 4. Consider cosine distance range normalization
> 
> ### ✅ Strengths:
> * Excellent separation of concerns
> * Proper encapsulation (SIMD internals are `pub(crate)`)
> * No breaking changes to public API
> * Comprehensive test coverage
> * Outstanding documentation
> * Follows all coding standards
> 
> **Recommendation:** Approve after fixing the critical magnitude comparison bug. Great work on this refactoring! 🎉
> 
> _Reviewed with adherence to [CLAUDE.md](../blob/trunk/CLAUDE.md) and [CODING_STANDARDS.md](../blob/trunk/docs/CODING_STANDARDS.md)_

@jules
